Even after #7429 merged, a crash AttributeError: 'ClassDef' object has no attribute 'value'

### Bug description

When parsing the following file, Pylint crashes:

```python
class AnotherClass:
    ...


class Pylint7429:
    def foo(self):
        self.__class__, myvar = AnotherClass, "myvalue"
```

Pylint versions prior to 2.9 run fine on the same file.

### Configuration

_No response_

### Command used

```shell
pylint /tmp/pylint7429.py
```


### Pylint output

```python
************* Module pylint7429
[...]
Exception on node <AssignAttr.__class__ l.7 at 0x7fb5f19060d0> in file '/tmp/pylint7429.py'
Traceback (most recent call last):
  File "pylint/utils/ast_walker.py", line 90, in walk
    callback(astroid)
  File "pylint/checkers/classes/class_checker.py", line 1556, in visit_assignattr
    self._check_invalid_class_object(node)
  File "pylint/checkers/classes/class_checker.py", line 1561, in _check_invalid_class_object
    inferred = safe_infer(node.parent.value)
  File "astroid/bases.py", line 134, in __getattr__
    return getattr(self._proxied, name)
AttributeError: 'ClassDef' object has no attribute 'value'
/tmp/pylint7429.py:1:0: F0002: /tmp/pylint7429.py: Fatal error while checking '/tmp/pylint7429.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/home/akaihola/.cache/pylint/pylint-crash-2022-09-15-14-06-01.txt'. (astroid-error)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```


### Expected behavior

Expect pylint to not crash


### Pylint version

```shell
pylint 2.15.2
astroid 2.12.9
Python 3.8.13 (default, Mar 16 2022, 13:02:57) 
[GCC 11.3.0]
```


### OS / Environment

NixOS 22.05

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/7467.bugfix
+++ b/doc/whatsnew/fragments/7467.bugfix
@@ -1,0 +1,3 @@
+``invalid-class-object`` does not crash anymore when ``__class__`` is assigned alongside another variable.
+
+Closes #7467

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -684,7 +684,7 @@ MSGS: dict[str, MessageDefinitionTuple] = {
         "Used when a value in __slots__ conflicts with a class variable, property or method.",
     ),
     "E0243": (
-        "Invalid __class__ object",
+        "Invalid assignment to '__class__'. Should be a class definition but got a '%s'",
         "invalid-class-object",
         "Used when an invalid object is assigned to a __class__ property. "
         "Only a class is permitted.",
@@ -1579,7 +1579,12 @@ a metaclass class method.",
         ):
             # If is uninferable, we allow it to prevent false positives
             return
-        self.add_message("invalid-class-object", node=node, confidence=INFERENCE)
+        self.add_message(
+            "invalid-class-object",
+            node=node,
+            args=inferred.__class__.__name__,
+            confidence=INFERENCE,
+        )
 
     def _check_in_slots(self, node: nodes.AssignAttr) -> None:
         """Check that the given AssignAttr node

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -1579,7 +1579,7 @@ a metaclass class method.",
         ):
             # If is uninferable, we allow it to prevent false positives
             return
-        self.add_message("invalid-class-object", node=node)
+        self.add_message("invalid-class-object", node=node, confidence=INFERENCE)
 
     def _check_in_slots(self, node: nodes.AssignAttr) -> None:
         """Check that the given AssignAttr node

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -1560,7 +1560,18 @@ a metaclass class method.",
     def _check_invalid_class_object(self, node: nodes.AssignAttr) -> None:
         if not node.attrname == "__class__":
             return
-        inferred = safe_infer(node.parent.value)
+        if isinstance(node.parent, nodes.Tuple):
+            class_index = -1
+            for i, elt in enumerate(node.parent.elts):
+                if hasattr(elt, "attrname") and elt.attrname == "__class__":
+                    class_index = i
+            if class_index == -1:
+                # This should not happen because we checked that the node name
+                # is '__class__' earlier, but let's not be too confident here
+                return  # pragma: no cover
+            inferred = safe_infer(node.parent.parent.value.elts[class_index])
+        else:
+            inferred = safe_infer(node.parent.value)
         if (
             isinstance(inferred, nodes.ClassDef)
             or inferred is astroid.Uninferable

--- a/tests/functional/i/invalid/invalid_class_object.py
+++ b/tests/functional/i/invalid/invalid_class_object.py
@@ -1,11 +1,14 @@
 # pylint: disable=missing-docstring,too-few-public-methods,invalid-name
 from collections import defaultdict
 
+
 class A:
     pass
 
+
 class B:
     pass
+
 
 A.__class__ = B
 A.__class__ = str
@@ -30,3 +33,43 @@ class C:
         obj = C()
         obj.__class__ = self.__class__
         return obj
+
+
+class AnotherClass:
+    ...
+
+
+class Pylint7429Good:
+    """See https://github.com/PyCQA/pylint/issues/7467"""
+
+    def class_defining_function_good(self):
+        self.__class__, myvar = AnotherClass, "myvalue"
+        print(myvar)
+
+    def class_defining_function_bad(self):
+        self.__class__, myvar = 1, "myvalue"  # [invalid-class-object]
+        print(myvar)
+
+    def class_defining_function_good_inverted(self):
+        myvar, self.__class__ = "myvalue", AnotherClass
+        print(myvar)
+
+    def class_defining_function_bad_inverted(self):
+        myvar, self.__class__ = "myvalue", 1  # [invalid-class-object]
+        print(myvar)
+
+    def class_defining_function_complex_bad(self):
+        myvar, self.__class__, other = (  # [invalid-class-object]
+            "myvalue",
+            1,
+            "othervalue",
+        )
+        print(myvar, other)
+
+    def class_defining_function_complex_good(self):
+        myvar, self.__class__, other = (
+            "myvalue",
+            str,
+            "othervalue",
+        )
+        print(myvar, other)

--- a/tests/functional/i/invalid/invalid_class_object.txt
+++ b/tests/functional/i/invalid/invalid_class_object.txt
@@ -1,5 +1,5 @@
-invalid-class-object:20:0:20:11::Invalid __class__ object:UNDEFINED
-invalid-class-object:21:0:21:11::Invalid __class__ object:UNDEFINED
-invalid-class-object:50:8:50:22:Pylint7429Good.class_defining_function_bad:Invalid __class__ object:UNDEFINED
-invalid-class-object:58:15:58:29:Pylint7429Good.class_defining_function_bad_inverted:Invalid __class__ object:UNDEFINED
-invalid-class-object:62:15:62:29:Pylint7429Good.class_defining_function_complex_bad:Invalid __class__ object:UNDEFINED
+invalid-class-object:20:0:20:11::Invalid __class__ object:INFERENCE
+invalid-class-object:21:0:21:11::Invalid __class__ object:INFERENCE
+invalid-class-object:50:8:50:22:Pylint7429Good.class_defining_function_bad:Invalid __class__ object:INFERENCE
+invalid-class-object:58:15:58:29:Pylint7429Good.class_defining_function_bad_inverted:Invalid __class__ object:INFERENCE
+invalid-class-object:62:15:62:29:Pylint7429Good.class_defining_function_complex_bad:Invalid __class__ object:INFERENCE

--- a/tests/functional/i/invalid/invalid_class_object.txt
+++ b/tests/functional/i/invalid/invalid_class_object.txt
@@ -1,2 +1,5 @@
-invalid-class-object:17:0:17:11::Invalid __class__ object:UNDEFINED
-invalid-class-object:18:0:18:11::Invalid __class__ object:UNDEFINED
+invalid-class-object:20:0:20:11::Invalid __class__ object:UNDEFINED
+invalid-class-object:21:0:21:11::Invalid __class__ object:UNDEFINED
+invalid-class-object:50:8:50:22:Pylint7429Good.class_defining_function_bad:Invalid __class__ object:UNDEFINED
+invalid-class-object:58:15:58:29:Pylint7429Good.class_defining_function_bad_inverted:Invalid __class__ object:UNDEFINED
+invalid-class-object:62:15:62:29:Pylint7429Good.class_defining_function_complex_bad:Invalid __class__ object:UNDEFINED

--- a/tests/functional/i/invalid/invalid_class_object.txt
+++ b/tests/functional/i/invalid/invalid_class_object.txt
@@ -1,5 +1,5 @@
-invalid-class-object:20:0:20:11::Invalid __class__ object:INFERENCE
-invalid-class-object:21:0:21:11::Invalid __class__ object:INFERENCE
-invalid-class-object:50:8:50:22:Pylint7429Good.class_defining_function_bad:Invalid __class__ object:INFERENCE
-invalid-class-object:58:15:58:29:Pylint7429Good.class_defining_function_bad_inverted:Invalid __class__ object:INFERENCE
-invalid-class-object:62:15:62:29:Pylint7429Good.class_defining_function_complex_bad:Invalid __class__ object:INFERENCE
+invalid-class-object:20:0:20:11::Invalid assignment to '__class__'. Should be a class definition but got a 'Instance':INFERENCE
+invalid-class-object:21:0:21:11::Invalid assignment to '__class__'. Should be a class definition but got a 'Const':INFERENCE
+invalid-class-object:50:8:50:22:Pylint7429Good.class_defining_function_bad:Invalid assignment to '__class__'. Should be a class definition but got a 'Const':INFERENCE
+invalid-class-object:58:15:58:29:Pylint7429Good.class_defining_function_bad_inverted:Invalid assignment to '__class__'. Should be a class definition but got a 'Const':INFERENCE
+invalid-class-object:62:15:62:29:Pylint7429Good.class_defining_function_complex_bad:Invalid assignment to '__class__'. Should be a class definition but got a 'Const':INFERENCE

